### PR TITLE
feat(dbAuth): Refactor dbAuthHandler to support WebAPI Request events

### PR DESCRIPTION
**Pull request by dac09**
### Context

I'm attampting to break down the SSR-Auth work into smaller, more managable, pieces.

This PR refactors mainly the `dbAuthHandler` to support both Lambda events and WebAPI Request. Key things to note:
- this PR on it's own does not enable SSR-Auth
- changes are non breaking, in SPA mode dbAuth should continue to work as normal
- Fetch/WebAPI support is incomplete in this PR, because in order to complete it we will need to make breaking changes. However with LambdaRequest it should continue to work (as if nothing changed)
- it **does** remove the extra check that compares that the userId in the cookie matches the one sent into the token. I will discuss the implications of this with @cannikin ✅ Cannikin approved
- The dbAuthHandler does support Fetch API Request, but..... lambda's base64 encoded body still needs to be handled _outside_ - see https://github.com/ardatan/whatwg-node/tree/master/packages/server#aws-lambda

~**However** this does break auth provider impersonation in RW studio - which is technically still experimental - so we're ok to do. See issue here: https://github.com/redwoodjs/redwood/issues/9801~

### Todos

- [x] Fix outstanding unit tests
- [x] Review edgecases in test project with Rob
- [x] Figure out what we're doing about auth impersonation

### CODE REVIEW

File: packages/auth-providers/dbAuth/api/src/shared.ts [hunk 1]
```diff
@@ -178,25 +190,36 @@ export const getSession = (
...
+  const bearerToken = extractEncryptedSessionFromHeader(event)
+
+  if (sessionCookie) {
+    // i.e. Browser making a request
     const [session, _csrfToken] = decryptSession(
-      getSession(extractCookie(event), cookieNameOption)
+      getSession(sessionCookie, cookieNameOption)
     )
+    return session
+  } else if (bearerToken) {
```
Comment (dac09) [inline packages/auth-providers/dbAuth/api/src/shared.ts:205]: @cannikin not sure about this one here. I don't think we need it right? This was for our first implementation of dbAuth ssr

File: packages/auth-providers/dbAuth/api/src/shared.ts [hunk 2]
```diff
@@ -90,8 +93,17 @@ const legacyDecryptSession = (encryptedText: string) => {
 
 // Extracts the session cookie from an event, handling both
 // development environment GraphiQL headers and production environment headers.
-export const extractCookie = (event: APIGatewayProxyEvent) => {
-  return eventGraphiQLHeadersCookie(event) || eventHeadersCookie(event)
+export const extractCookie = (event: APIGatewayProxyEvent | Request) => {
+  // @TODO Disabling Studio Auth impersonation: it uses body instead of headers
```
Comment (dac09) [inline packages/auth-providers/dbAuth/api/src/shared.ts:97]: Remove TODO

File: packages/auth-providers/dbAuth/api/src/decoder.ts [hunk 3]
```diff
@@ -5,28 +5,24 @@ import type { Decoder } from '@redwoodjs/api'
...
     const session = dbAuthSession(req.event, cookieNameOption)
-    const authHeaderUserId = token
-
-    if (session.id.toString() !== authHeaderUserId) {
-      console.error('Authorization header does not match decrypted user ID')
-      throw new Error('Authorization header does not match decrypted user ID')
-    }
 
+    // We no longer compare the session id with the bearer token
+    // Because we only pass around the encrypted session (in both cookie and header)
```
Comment (dac09) [inline packages/auth-providers/dbAuth/api/src/decoder.ts:16]:
```suggestion
```

File: packages/auth-providers/dbAuth/api/src/__tests__/shared.test.ts [hunk 4]
```diff
@@ -320,7 +322,9 @@ describe('session cookie extraction', () => {
       expect(extractCookie(event)).toEqual(cookie)
     })
 
-    it('overwrites cookie with event header GraphiQL when in dev', () => {
+    // @TODO: Disabled Studio Auth Implementation
+    // we need to avoid using body instead of headers
+    it.skip('overwrites cookie with event header GraphiQL when in dev', () => {
```
Comment (dac09) [inline packages/auth-providers/dbAuth/api/src/__tests__/shared.test.ts:327]: UNDO SKIP!

File: packages/auth-providers/dbAuth/web/src/dbAuth.ts [hunk 5]
```diff
@@ -86,6 +86,10 @@ export function createDbAuthClient({
       return getTokenPromise
     }
 
+    // Set-Cookie: same-session-xxx-yy
```
Comment (dac09) [inline packages/auth-providers/dbAuth/web/src/dbAuth.ts:89]: Remove comment here

File: packages/auth-providers/dbAuth/api/src/shared.ts (hunk 1 above)
Comment (cannikin) [reply packages/auth-providers/dbAuth/api/src/shared.ts:205]: Yeah if you're just adding the cookie itself to the headers then we don't need this auth header bearer token check any more!

Comment (dac09) [reply packages/auth-providers/dbAuth/api/src/shared.ts]: Thanks, taken out. Will just be forwarding on the cookie in the SSR-Auth version

**Comment by dac09** [comment]
Ready for final review @cannikin! Thanks ✌️

**Review by cannikin — CHANGES_REQUESTED** [review]
We talked about moving all of the `await init()` calls to the test file, since that seems to be the only place they're needed (except for `invoke()`).

**Comment by dac09** [comment]
Dankeschone @cannikin ✌️

**Pull request merged**